### PR TITLE
Issue 790. Add gocean grid properties intrinsic type

### PR DESCRIPTION
**Pull request by sergisiso**
This PR fixes one of the issues with generating OpenCL for the Shallow benchmark which is the lack of type information on the gocean grid properties, this has been added as a new configuration field.

Additionally it adds a `runme_opencl.py` script in `examples/gocean/eg1` which is the example that resembles the Shallow benchmark.

### CODE REVIEW

**Comment by sergisiso** [comment]
@arporter @hiker As discussed in issue #790 I added a new field to provide the intrinsic size of the gocean grid_properties in the config file. This is ready for review.

File: src/psyclone/configuration.py [hunk 1]
```diff
@@ -937,20 +937,21 @@ def __init__(self, config, section):
...
                 all_props = self.create_dict_from_string(section[key])
                 for grid_property in all_props:
                     try:
-                        fortran, variable_type = all_props[grid_property]\
-                                            .split(":")
+                        fortran, variable_type, intrinsic_type = \
+                            all_props[grid_property].split(":")
                     except ValueError:
                         # Raised when the string does not contain exactly
                         # two values separated by ":"
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/configuration.py:951]: Done

File: src/psyclone/configuration.py [hunk 2]
```diff
@@ -937,20 +937,21 @@ def __init__(self, config, section):
...
                         error = "Invalid property \"{0}\" found with value " \
                                 "\"{1}\" in \"{2}\". It must have exactly " \
-                                "two ':'-delimited separated values: the " \
+                                "three ':'-delimited separated values: the " \
                                 "property and the type." \
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/configuration.py:955]: Done

File: src/psyclone/configuration.py [hunk 3]
```diff
@@ -995,26 +997,34 @@ def make_property(dereference_format, type_name):
...
         '''
         if type_name not in ['array', 'scalar']:
             from psyclone.errors import InternalError
             raise InternalError("Type must be 'array' or 'scalar' but is "
                                 "'{0}'.".format(type_name))
 
-        Property = namedtuple("Property", "fortran type")
-        return Property(dereference_format, type_name)
+        if intrinsic_type not in ['integer', 'real']:
+            from psyclone.errors import InternalError
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/configuration.py:1015]: Done

File: src/psyclone/tests/gocean1p0_opencl_test.py [hunk 4]
```diff
@@ -299,6 +299,32 @@ def test_set_kern_args(kernel_outputdir):
...
 
+def test_set_kern_args_real_grid_property():
+    ''' Check that we generate correct code to set a real scalar grid
+    property. '''
+    psy, _ = get_invoke("driver_test.f90", API, idx=0)
+    sched = psy.invokes.invoke_list[0].schedule
+    otrans = OCLTrans()
+    otrans.apply(sched)
+    generated_code = str(psy.gen)
+    print(generated_code)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/gocean1p0_opencl_test.py:310]: Done

File: examples/gocean/eg1/Makefile [hunk 5]
```diff
@@ -57,3 +57,6 @@ dag:
 
 openacc:
 	${PYTHON} ./runme_openacc.py
+
+opencl:
+	${PYTHON} ./runme_opencl.py
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/gocean/eg1/Makefile:62]: Yes the psyclone invocation looks better.

File: examples/gocean/eg1/runme_opencl.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: S. Siso, STFC Daresbury Lab
+
+'''A simple test script showing the introduction of OpenCL with PSyclone.
+In order to use it you must first install PSyclone like so:
+
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/gocean/eg1/runme_opencl.py]: Done

**Comment by sergisiso** [comment]
@arporter I made the suggested modifications, this is ready for next review

File: examples/gocean/eg1/README.md [hunk 7]
```diff
@@ -16,6 +16,17 @@ python ./runme.py
...
 Examine the runme*.py scripts themselves for further details.
 
+
+## OpenCL PSyclone script
+
+The OpenCL transformation is provided with a PSyclone transformation script
+(`opencl_transformation.py`). This can be run using the PSyclone command:
+
+```sh
+psyclone -s ./opencl_transformation.py -api gocean1.0 shallow_alg.f90
```
Comment (arporter) [inline examples/gocean/eg1/README.md:26]: Do you need the `--kernel-renaming` flag here?

File: examples/gocean/eg1/opencl_transformation.py [hunk 8]
```diff
@@ -33,50 +33,42 @@
 # -----------------------------------------------------------------------------
 # Authors: S. Siso, STFC Daresbury Lab
 
-'''A simple test script showing the introduction of OpenCL with PSyclone.
-In order to use it you must first install PSyclone like so:
+''' Module providing a transformation script that converts the Schedule of
```
Comment (arporter) [inline examples/gocean/eg1/opencl_transformation.py:36]: s/a transformation script/a PSyclone transformation script/

File: examples/gocean/eg1/opencl_transformation.py [hunk 9]
```diff
@@ -33,50 +33,42 @@
...
+    each Invoke to use OpenCL. '''
 
- >>> pip install --user psyclone
 
-(or see the Getting Going section in ../../psyclone.pdf.) Once PSyclone is
-installed, this script may be run by doing:
+def trans(psy):
+    '''
+    Transformation routine for use with PSyclone. Applies the OpenCL
+    transform.
```
Comment (arporter) [inline examples/gocean/eg1/opencl_transformation.py:43]: "Converts any global-variable accesses into kernel arguments and then applies the OpenCL transformation to the PSy layer"

File: examples/gocean/eg1/opencl_transformation.py [hunk 10]
```diff
@@ -33,50 +33,42 @@
...
-from psyclone.psyGen import PSyFactory, TransInfo
+    for invoke in psy.invokes.invoke_list:
+        print("Converting to OpenCL invoke: " + invoke.name)
+        schedule = invoke.schedule
 
-API = "gocean1.0"
-_, INVOKEINFO = parse("shallow_alg.f90", api=API)
-PSY = PSyFactory(API).create(INVOKEINFO)
-print(PSY.gen)
+        # Skip invoke_2
```
Comment (arporter) [inline examples/gocean/eg1/opencl_transformation.py:62]: Please could you extend the comment to say why we skip invoke_2.

File: examples/gocean/eg1/opencl_transformation.py (hunk 8 above)
Comment (sergisiso) [reply examples/gocean/eg1/opencl_transformation.py:36]: Done

File: examples/gocean/eg1/opencl_transformation.py (hunk 9 above)
Comment (sergisiso) [reply examples/gocean/eg1/opencl_transformation.py:43]: Done

File: examples/gocean/eg1/opencl_transformation.py (hunk 10 above)
Comment (sergisiso) [reply examples/gocean/eg1/opencl_transformation.py:62]: Extended comment and referenced an issue.

File: examples/gocean/eg1/README.md (hunk 7 above)
Comment (sergisiso) [reply examples/gocean/eg1/README.md:26]: We do if we were to compile the generated OpenCL. I added a comment in the Makefile explaining why we need it.

**Comment by sergisiso** [comment]
@arporter Addressed your comments, PR ready for next review.

**Comment by arporter** [comment]
Both Travis runs are green, despite the fact that GitHub says the checks haven't completed.

**Pull request merged**